### ISSUE TITLE
MachO support

### ISSUE THREAD

**Issue by jrmuizel** [yegord/snowman#8]
This would be nice to have.

**Comment by yegord** [comment]
You are more than welcome to start it.
You can start with writing an architecture class, then a disassembler.
Then, what will be left are description of the semantics of instructions, and description(s) of calling convention(s).

**Comment by archshift** [comment]
Mach-O is not a separate, architecture, just an executable format like ELF or PE. It's used on Macs. Is all that really necessary?

**Comment by yegord** [comment]
Sorry, it's late, I am doing several things in parallel, and I am writing nonsense. :-) Yes, one needs only to write a parser. :)

**Comment by yegord** [comment]
It would be great to delegate parsing to a thirty-party library, like libbfd. Last time I investigated this option it was not clear for me how to properly integrate libbfd into our CMake-based build process. It was also unclear, whether libbfd builds fine with MSVC 2010 and later. It would be great if somebody would investigate these things.

BTW, what do other projects use for parsing object files? Do they invent a bicycle every time?

**Comment by jrmuizel** [comment]
I'd recommend against using libbfd. It's not a great code base and doesn't necessarily work that well for object formats other than elf. LLVM has object parsers that I think would work.

- radare uses it's own parsers.
- mc-sema uses a it's own parser for PE, and LLVM for COFF and ELF
- fracture uses LLVM

**Comment by yegord** [comment]
LLVMObject looks nice, and should be buildable by all the compilers we need. (Not by MSVC 2010, but we need it only for the IDA plug-in, where we do not need parsers.)

Any good ideas on how to avoid making the users build LLVM separately (e.g., on Windows)? Keep the necessary part in our source tree?

**Comment by jrmuizel** [comment]
It looks LLVMObject only depends on LLVMSupport and already has a CMake based build system. It seems like it would work to just rip the Object and Support directories out of LLVM and include them in the snowman source tree. That would avoid having an external dependency on LLVM (which would be a  pain) and isn't an enormous amount of source code which should keep the repo size and build times down.

**Comment by nihilus** [comment]
Should not be to harsh to re-implement the code from boomerang.sf.net :-)

**Comment by jrmuizel** [comment]
I have a prototype that extracts LLVMObject and the required pieces of LLVMSupport from LLVM and links them into snowman. It's not beautiful, but wasn't extremely ugly either (I only needed to patch one source file)

Building all of LLVMSupport seemed to require more build infrastructure and headers so I chose not to take that route.

**Comment by yegord** [comment]
Maybe it makes sense first to try to write a parser using external LLVM library. LLVM D Compiler has a reasonably written FindLLVM.cmake.

**Comment by nihilus** [comment]
Well O think this might up in a dead end... The boomerang project had many flaws. One of the got rid of by using capstone another one would have been to use libbfd instead of separate parsers.

**Comment by nihilus** [comment]
As far as it goes for the quality of libbfd, I recall it was invented by the ingenious people at Cygnus (they also made cygwin and newlib before RedHat bought them) of which a few Ive got to meet and know, notably Rob Savoye. I can say that I dont mind checking up how to integrate libbfd.

**Comment by hlide** [comment]
for mach-o: https://github.com/vertis/objconv

**Comment by hlide** [comment]
By the way, it also contains a disassembler for x86[-64] which can even disassemble AVX3 (AVX-512).

**Comment by nihilus** [comment]
@yegord shouldn't this issue be closed now? Or do we need to test the Mach-O parser?

**Comment by nihilus** [comment]
I can tell it is decompiling it self for me: file snowman/snowman 
snowman/snowman: Mach-O 64-bit x86_64 executable

**Comment by yegord** [comment]
I still need to implement parsing of dysymtab and add regression tests, so, let's keep this issue open for a while.

**Comment by nihilus** [comment]
@yegord What was the issue you had with libbfd? Converting it to CMake and then integrating it?

**Comment by yegord** [comment]
@nihilus Yes.

**Comment by nihilus** [comment]
However for Mach-O I recommend checking this out: https://github.com/nemerle/boomerang/tree/next/loader/machO

**Comment by nihilus** [comment]
I had a rain check at libbfd during work today... and yes, making it build with CMake should be doable. The big thing is the integration.

**Comment by yegord** [comment]
Snowman already has a parser for Mach-O, and this parser seems to work.
I do not plan to replace it with something else.
If someone wants it to be replaced with some other parser, one has to
1) implement the other parser,
2) make it compile without warnings and work under all supported compilers, systems, and architectures,
3) submit a pull request,
4) convince me that the new parser is better than what there is now.

**Comment by nihilus** [comment]
@yegord I told you it works good... However dysymtab you could find inspiration there or for other formats :-)

**Comment by yegord** [comment]
Okay, actually, there is nothing interesting in dysymtab.

**Comment by jrmuizel** [comment]
dysymtab should be useful for getting imported functions.
i.e. if a MachO shared library calls 'pow' there will be an entry in the dysymtab pointing at pow. This is useful information to have in the decompiler output.

**Comment by nihilus** [comment]
@jrmuizel nice info!

**Comment by jrmuizel** [comment]
I have this basically working here:
https://github.com/jrmuizel/snowman/tree/dysymtab

**Comment by yegord** [comment]
Do I understand correctly that the main point of parsing dysymtab is to handle indirect symbols properly? If this is true, we should probably not add these symbols at https://github.com/jrmuizel/snowman/commit/395425e73d99818df80024a85723ec9933d013c0#diff-d1352485d472009efda02a15ec399107R332
